initial support for adc buttons

mcmatrix opened the pull request:
Development for #898
This is **work in progress**
Tried to make initial support for ADC buttons.

In `[display]` section you can specify pins for the menu.
Now it supports the following format for analog (ADC) buttons.
`up_pin: analog15(100,200)`
`up_pin: pin_name(min, max)`
`min` and `max` are concrete button adc value range.
These are always integers.

The max ADC range can be set `adc_range: 1024` attribute (default is 1024).
Internally klipper has ADC value range from 0.0 to 1.0 and this will convert the value to an integer. 

Only these pins allow ADC input
```ini
#click_pin:
#   The pin connected to 'enter' button or encoder 'click'. This parameter
#   must be provided when using menu.
#back_pin:
#   The pin connected to 'back' button. This parameter is optional, menu
#   can be used without it.
#up_pin:
#   The pin connected to 'up' button. This parameter must be provided
#   when using menu without encoder.
#down_pin:
#   The pin connected to 'down' button. This parameter must be provided
#   when using menu without encoder.
#kill_pin:
#   The pin connected to 'kill' button. This button will call
#   emergency stop.
```

It's not fully tested. 

@KevinOConnor Could you check `buttons.py` ADC part.
Maybe you have some comments or suggestions.

Signed-off-by: Janar Sööt <janar.soot@gmail.com>

mcmatrix commented:
Now I have to setup buttons and resistors on the breadboard.
![image](https://user-images.githubusercontent.com/1433580/49688244-bd16ae80-fb17-11e8-82a0-0960af79b5fb.png)
To test it correctly.

KevinOConnor commented:
Thanks.  My very high-level comments would be:
1 - I'd recommend moving adc_callback() into a new MCU_ADC_buttons class (similar to the way the MCU_buttons code does that).
2 - The code will need to use the reactor.register_async_callback() mechanism as the adc_callback() is invoked in a high-priority message parsing thread while the main code runs in a different thread.  See how MCU_buttons does this.
3 - I'd prefer to avoid syntax parsing of the config fields.  I recommend being explicit with something like `analog_up_pin: analog15`, `analog_min_up_pin: 100`, `analog_max_up_pin: 200`.  Or, if you prefer just `analog_up_pin: analog15` and `analog_range_up_pin: 100, 200`.

Thanks again,
-Kevin

rmcbc commented:
Hello @mcmatrix are you going make the changes that kevin suggested?
I try to use this code with a LCD2004 from a Anet A8 and don't work correctly.
If you want I can try to solve this.

mcmatrix commented:
@rmcbc Thanks for your help. At the moment I hope to handle this.
Yes, I'll implement Kevin's suggestions and will improve the functionality of the buttons.
This PR is not reflecting my latest local changes, I'll update this PR when it's more or less working
in my workbench.

mcmatrix commented:
2nd attempt.
Remade buttons functionality as Kevin suggested.
Added better debounce for ADC buttons.
Added option `adc_debug:` Default is False (disabled)
When True (enabled) it outputs scaled ADC value to log.

ADC pin definition in the config is still old way `up_pin: pin_name(min, max)`

mcmatrix commented:
@rmcbc Thank you for the feedback.
How quickly are your buttons working?
On my bench having a breadboard with buttons, it'll skip
when clicking too quickly. I have to play around with ADC reporting
time, it's also a debounce time.
In `buttons.py`
`ADC_REPORT_TIME = 0.050`
Currently it's 50ms and it makes 100ms for debounce checking.
I have to check will the quicker timing make things better.

ZhenyaKa commented:
> adc_debug

What section for adc_debug?

ZhenyaKa commented:
WOW It's works! Some delayed but works!

My config with debug values. Right and Center button do the same action.

# "RepRapDiscount 2004 Smart Controller" type displays
[display]
#adc_debug: true
lcd_type: hd44780
rs_pin: ar41
e_pin: ar47
d4_pin: ar43
d5_pin: ar37
d6_pin: ar45
d7_pin: ar39
#encoder_pins: ^ar31, ^ar33
#click_pin: ^!ar35
#up 672
#down 173
#left 90
#right 494
#center 315
up_pin:   analog12(640,700)
down_pin: analog12(140,200)
back_pin: analog12(60,120)
#right_pin:analog12(460,530)
click_pin:analog12(280,530)

mcmatrix commented:
> WOW It's works! Some delayed but works!

Thanks. What you mean by delayed?

ZhenyaKa commented:
> > WOW It's works! Some delayed but works!
> 
> Thanks. What you mean by delayed?

not so fast response like was before (Repetier FW)

derbengale commented:
I had another issue: when I start Klipper using sudo service klipper start then it doesnt start kipper anymore. The service is started but klippy doesnt see to start, also /tmp/printer doesnt show up anymore. I have to start Klipper using

source klippy-env/bin/activate
and then 
python klipper/klippy/klippy.py printer.cfg

Is someone else having this issue?

mcmatrix commented:
> had another issue: when I start Klipper using sudo service klipper start then it doesnt start kipper anymore. 

@derbengale What the log file is saying?

derbengale commented:
It does not exist. That's why I assume klippy doesn't start. When I start the service, where is the path to my klippy defined? Maybe I should check there

mcmatrix commented:
@rmcbc
> But it works ok, a bit slow but later I'm going to change ADC_REPORT_TIME to find a better value.

Have you had a chance to test `ADC_REPORT_TIME` value change?

mcmatrix commented:
> Separately, as just a random idea, we could make the presence of an analog_range_click_pin config parameter turn the existing click_pin parameter from digital to analog. If you feel that's an improvement.

Thanks. I think it's a good idea, I'll implement it.

mcmatrix commented:
Hope to update this in a few days

mcmatrix commented:
Updated version!
Reduced ADC reporting time from 50ms to 30ms, it should reduce debounce time.
I hope it has a positive effect.

**NB! Analog pins config has been changed!**
Implemented Kevin's idea about pin attributes.
Please look `example-extras.cfg` for additonal info.

KevinOConnor commented:
Thanks!  I suspect we'll need to merge to get more feedback.  Let me know when you'd like me to do this.

One comment on the parameters - I wonder if there is a way to use meaningful units instead of the analog_pin_scale parameter.  Scaling to 1024 seems like an odd thing to do.

-Kevin

mcmatrix commented:
> One comment on the parameters - I wonder if there is a way to use meaningful units instead of the analog_pin_scale parameter. Scaling to 1024 seems like an odd thing to do.

For me it has meaning, I'm coming from 10bit Arduino ADC :)
But yes, you have point.

Maybe something from followings:
Decimal voltage: `0 - 500`, `0 - 330` etc
Float voltage: `0.00 - 5.00`, `0.00 - 3.30 `etc with 2 decimal places
or Percentage `0 - 100` or `0.0 - 100.0` with one decimal place
Per thousand `0 - 1000` or something else... any ideas.
It depends how much precision we need?

KevinOConnor commented:
FWIW, my recommendation would be to allow the user to specify the pull_up resistance, and then allow them to specify the button resistance range.  (The formula for determining resistance is `r = pullup * adc / (1.0 - adc)` ).  The pull_up resistance can default to 4700 ohms if that is common.

Separately, I'm fine with merging the code as is.

-Kevin

mcmatrix commented:
Thanks. I'll check this pull_up resistance approach.
Let you know how it goes.

mcmatrix commented:
@KevinOConnor 
I changed it to resistance range by using that pullup formula.
I had to reduce ADC resolution (round ADC values) to make buttons work better.
But it still cannot detect quick press.

Maybe it's that ADC reporting system. It's good for sensors and temperature
but for the analogue button, it's not right. For me, it looks that the reported ADC running value
takes time when it reaches to the real value.
I don't know how to make it better, maybe in MCU level but I don't have time to do it.
Maybe somebody else has skills and idea has to improve this.

Analogue keypad users, please test it. If it's more or less okay we can put it into live
until somebody will make it better.

KevinOConnor commented:
Thanks.  If you think it is ready, I think we should merge this.  I suspect more feedback will come if it is in the master branch.

-Kevin

derbengale commented:
@mcmatrix I just tried to use it but now I can't manage to get klipper working because im obviously doing something wrong. Octoprint says 

> Recv: // Option 'analog_up_pin' is not valid in section 'display'Recv: // Once the underlying issue is >corrected, use the "RESTART"Recv: // command to reload the config and restart the host software.Recv: >// Printer is haltedRecv: // Klipper state: Not readyRecv: !! Option 'analog_up_pin' is not valid in section >'display'

I used your example in my printer.cfg
`analog_up_pin: PA1
analog_range_up_pin: 480,520
analog_down_pin: PA1
analog_range_down_pin: 680,720
analog_click_pin: PA1
analog_range_click_pin: 160,190`

I now removed all the lines with my pin definition PA1 but now the buttons dont work anymore of course. Where do I define my analog pin?

mcmatrix commented:
> How do I use the debug feature you built in to give you some proper information you can work with?

The range unit has been changed. It's not plain ADC value anymore.
It's the resistance. 
You'll have to set pull-up resistor value if it's different from 4700 ohms.
And then resistance range for each button. You have to debug those values again (ADC1024 values will not work).
```ini
#analog_pullup_resistor: 4700
#   The resistance (in ohms) of the pullup attached to the analog button.
#   The default is 4700 ohms.
#analog_pin_debug:
#   When enabled it will output analog (ADC) button readings to the log.
#   It's useful for finding analog button resistance range values.
#   The default is False (disabled)
#analog_range_click_pin:
#   The resistance range for a 'enter' button. Range minimum and maximum
#   comma-separated values must be provided when using analog button.
```

I hope it helps. Let me know how it goes?

derbengale commented:
Unfortunately the resistor value of 4,7kOhms is the resistor value from anets display board. I guess that picture above also came from the initial post about anets display as well. Is the input logged somewhere? Also I didn't see any point in the latest configuration where I could set up the adc pin. Is it maybe because of that?

mcmatrix commented:
> Edit: okay, I think I got it now. So I have to activate the debug first, then click all buttons then read the log and then set the range according to the log.

Yep, more or less the same as you did for ADC values when you didn't know the values.
The `pin name` attribute is the same for both digital and analogue.
Then you set `analog_range_xxx` for pin then the system assumes that it's analogue pin input.

KevinOConnor commented:
FYI, it seems users of the anet a8 would like to use this feature - for example, see: https://www.youtube.com/watch?v=EX8dJkdcmJQ

I suspect those users don't know about this PR.  If you think it is ready, we could commit it, update the printer-anet-a8-2017.cfg file and hope we get more testers.

-Kevin

mcmatrix commented:
I think we can commit it.
Its performance is not as good as the digital buttons work
but at the moment I don't know how to improve it.

Adrian-at-CrimsonAuzre commented:
Would this also be useful for replacing the FSR sensor board on the Voron @VoronDesign?

noud-github commented:
tested it. works great with anet a8 display

KevinOConnor commented:
Thanks.

@noud-github - could you post your anet a8 config?  (Or, if you are up for it, a pull request updating the example anet a8 config file.)

-Kevin

noud-github commented:
will do a pull request later this week, did the testing on a bigtreetech SKR, still have the anet stuff lying around so need to do some building first should look like this but not sure the range are te same 
```
up_pin: PA1
analog_range_up_pin: 470000,600000 
down_pin: PA1 
analog_range_down_pin: 3900,4100 
click_pin: PA1
analog_range_click_pin: 9000,11000

analog_pullup_resistor: 4700
#analog_pin_debug:true
```

ArDrift commented:
For me these values work on the Anet A8:
```
analog_range_up_pin: 9990,10050
analog_range_down_pin: 990,1000
analog_range_click_pin: 2150,4670
analog_range_back_pin: 460, 480
```

noud-github commented:
@ArDrift  thanx,  will use those as second opinion to my findings

aeropic commented:
Hi all,
I just tested the analog buttons on my July 2017 ANET A8.
I made a git pull, compiled the firmware, uploades it to the ANETs board, restarted Klippy, changed the config file to add the analog buttons.
==> If I leave all up_pin, down-pin ... I get an error saying pin PA1 is used several times in the config file
==> If I leave only the up_pin: PA1 and analog_range_up_pin: 9000, 13000, I get another error :
"Option 'analog_range_up_pin' is not valid in section 'display'"

For sure, I'm doing something wrong, but I cannot figure out what ????
Any help appreciated ;-)
Thanks
Alain
[printer.cfg.txt](https://github.com/KevinOConnor/klipper/files/3144491/printer.cfg.txt)

aeropic commented:
I tested the new ANET's menu button feature... It's working very well (with the default values in the generic ANETs file)
Thanks for having developed this !

doctor3lo commented:
Hi

Maybe somebody do this on
Ramps + anet LCD2004 + 5 buttons?

Show your config
Thank you.

duskyme commented:
> Hi
> 
> Maybe somebody do this on
> Ramps + anet LCD2004 + 5 buttons?
> 
> Show your config
> Thank you.

Anet A8 display connected to the AUX2 header. 
You must **switch the cables that are connected to the pin 1 and pin 2.  Pin 1 on AUX2 is VCC and pin 2 on AUX2 is GND, this is reversed on the Anet A8 display.**

Config
```
[display]
lcd_type: hd44780
rs_pin: ar64
e_pin: ar44
d4_pin: ar63
d5_pin: ar40
d6_pin: ar42
d7_pin: ar65
up_pin: analog12
analog_range_up_pin: 9000, 13000
down_pin: analog12
analog_range_down_pin: 800, 1300
click_pin: analog12
analog_range_click_pin: 2000, 2500
back_pin: analog12
analog_range_back_pin: 4500, 5000
```

gvnlws commented:
> > Hi
> > Maybe somebody do this on
> > Ramps + anet LCD2004 + 5 buttons?
> > Show your config
> > Thank you.
> 
> Anet A8 display connected to the AUX2 header.
> You must **switch the cables that are connected to the pin 1 and pin 2. Pin 1 on AUX2 is VCC and pin 2 on AUX2 is GND, this is reversed on the Anet A8 display.**
> 
> Config
> 
> ```
> [display]
> lcd_type: hd44780
> rs_pin: ar64
> e_pin: ar44
> d4_pin: ar63
> d5_pin: ar40
> d6_pin: ar42
> d7_pin: ar65
> up_pin: analog12
> analog_range_up_pin: 9000, 13000
> down_pin: analog12
> analog_range_down_pin: 800, 1300
> click_pin: analog12
> analog_range_click_pin: 2000, 2500
> back_pin: analog12
> analog_range_back_pin: 4500, 5000
> ```

Thank you so much!! I have been trying to work this out for a few hours and google wasn't being helpful. Works great for me :)
